File: 2024/week51-60/week56/414_17837.java
import java.io.*;
import java.util.*;

public class Main{
    static ArrayList<Integer>[][] board_hole;
    static int[][] piece;
    static int[] di={0,0,-1,1};
    static int[] dj={1,-1,0,0};
    static int result=0;
    static int k;
    public static boolean move_forward(int i,int j, int ni, int nj,int index, boolean reverse){
        int idx=board_hole[i][j].indexOf(index);
        List<Integer> remain=board_hole[i][j].subList(0,idx);
        List<Integer> moving=board_hole[i][j].subList(idx, board_hole[i][j].size());
        for (int l:moving) {
            piece[l][0]=ni;
            piece[l][1]=nj;
        }
        if(reverse){
            Collections.reverse(moving);
        }
        board_hole[ni][nj].addAll(moving);
        if(board_hole[ni][nj].size()>=4){
            System.out.println(result);
            return true;
        }else{
            board_hole[i][j]= new ArrayList(remain);
            return false;
        }
    }

    public static int change_dir(int index){
        if (piece[index][2]==1)
            piece[index][2]=0;
        else if (piece[index][2]==0)
            piece[index][2]=1;
        else if (piece[index][2]==2)
            piece[index][2]=3;
        else if (piece[index][2]==3)
            piece[index][2]=2;
        return piece[index][2];
    }

    public static void main(String[] args) throws IOException{
        BufferedReader br=new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st=new StringTokenizer(br.readLine());
        int n= Integer.parseInt(st.nextToken());
        k= Integer.parseInt(st.nextToken());
        int[][] board=new int[n][n];
        for (int i = 0; i < n; i++) {
            st=new StringTokenizer(br.readLine());
            for (int j = 0; j < n; j++) {
                board[i][j]=Integer.parseInt(st.nextToken());
            }
        }
        board_hole=new ArrayList[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                board_hole[i][j]=new ArrayList<Integer>();
            }
        }
        piece=new int[k][3];
        for (int i = 0; i < k; i++) {
            st=new StringTokenizer(br.readLine());
            int a=Integer.parseInt(st.nextToken())-1;
            int b=Integer.parseInt(st.nextToken())-1;
            int c=Integer.parseInt(st.nextToken())-1;
            board_hole[a][b].add(i);
            piece[i][0]=a;
            piece[i][1]=b;
            piece[i][2]=c;
        }
        boolean flag=false;
        while(result<=1000){
            result+=1;
            flag=false;
            for (int iii = 0; iii < k; iii++) {
                int i=piece[iii][0];
                int j=piece[iii][1];
                int d=piece[iii][2];
                int ni=i+di[d];
                int nj=j+dj[d];
                if(ni>=0 && ni<n && nj>=0 && nj<n){
                    if(board[ni][nj]==0){
                        flag=move_forward(i,j,ni,nj,iii,false);
                    } else if (board[ni][nj]==1) {
                        flag=move_forward(i,j,ni,nj,iii,true);
                    }
                    else{
                        int pre_dir=change_dir(iii);
                        ni=i+di[pre_dir];
                        nj=j+dj[pre_dir];
                        if (ni>=0 && ni<n && nj>=0 && nj<n && board[ni][nj]!=2){
                            if(board[ni][nj]==0){
                                flag=move_forward(i,j,ni,nj,iii,false);
                            }else{
                                flag=move_forward(i,j,ni,nj,iii,true);
                            }
                        }
                    }
                }else{
                    int pre_dir=change_dir(iii);
                    ni=i+di[pre_dir];
                    nj=j+dj[pre_dir];
                    if (ni>=0 && ni<n && nj>=0 && nj<n && board[ni][nj]!=2){
                        if(board[ni][nj]==0){
                            flag=move_forward(i,j,ni,nj,iii,false);
                        }else{
                            flag=move_forward(i,j,ni,nj,iii,true);
                        }
                    }
                }

                if(flag){
                    return;
                }
            }

        }
        System.out.println(-1);
        return;
    }
}

File: 2024/week51-60/week56/414_17837.py
import sys
input=sys.stdin.readline

n,k=map(int,input().split())
board=[list(map(int,input().split())) for _ in range(n)]
board_hole=[[[] for _ in range(n)] for _ in range(n)]
piece=[]
for i in range(k):
    a,b,c=map(int,input().split())
    a,b,c=a-1,b-1,c-1
    board_hole[a][b].append(i)
    piece.append([a,b,c])

di=[0,0,-1,1]
dj=[1,-1,0,0]
result=0
def move_forward(i,j,ni,nj,index,reverse=False):
    global result,k
    idx=board_hole[i][j].index(index)
    remain=board_hole[i][j][:idx]
    moving=board_hole[i][j][idx:]
    for l in moving:
        piece[l][0]=ni
        piece[l][1]=nj
    if reverse:
        moving.reverse()
    board_hole[ni][nj].extend(moving)
    if len(board_hole[ni][nj])>=4:
        print(result)
        exit()
    board_hole[i][j]=remain

while result<=1000:
    result+=1
    # for v in range(n):
    #     for b in range(n):
    #         temp_str=""
    #         for s in board_hole[v][b]:
    #             temp_str+=str(s)
    #             if piece[s][2]==0:
    #                 temp_str+=">"
    #             elif piece[s][2]==1:
    #                 temp_str+="<"
    #             elif piece[s][2]==2:
    #                 temp_str+="^"
    #             elif piece[s][2]==3:
    #                 temp_str+="-"
    #         print(board[v][b],end="")
    #         print("{0:>10}".format(temp_str),end=" | ")
    #     print("\n----------------------------------------------------------------------------")
    # print()
    # print("===================================================================================")
    # print()
    for index,value in enumerate(piece):
        i,j,d=value
        ni=i+di[d]
        nj=j+dj[d]
        if 0<=ni<n and 0<=nj<n:
            if board[ni][nj]==0:
                move_forward(i,j,ni,nj,index)
            elif board[ni][nj]==1:
                move_forward(i,j,ni,nj,index,reverse=True)
            else:
                if piece[index][2]==1:
                    piece[index][2]=0
                elif piece[index][2]==0:
                    piece[index][2]=1
                if piece[index][2]==2:
                    piece[index][2]=3
                elif piece[index][2]==3:
                    piece[index][2]=2

                ni=i+di[piece[index][2]]
                nj=j+dj[piece[index][2]]
                if 0<=ni<n and 0<=nj<n and board[ni][nj]!=2:
                    if board[ni][nj]==0:
                        move_forward(i,j,ni,nj,index)
                    else:
                        move_forward(i,j,ni,nj,index,reverse=True)
        else:
            if piece[index][2]==1:
                piece[index][2]=0
            elif piece[index][2]==0:
                piece[index][2]=1
            if piece[index][2]==2:
                piece[index][2]=3
            elif piece[index][2]==3:
                piece[index][2]=2
            ni=i+di[piece[index][2]]
            nj=j+dj[piece[index][2]]
            if 0<=ni<n and 0<=nj<n and board[ni][nj]!=2:
                if board[ni][nj]==0:
                    move_forward(i,j,ni,nj,index)
                else:
                    move_forward(i,j,ni,nj,index,reverse=True)

print(-1)
